brainglobe_registration/automated_target_selection.py

```diff
@@ -0,0 +1,392 @@
+    weights: tuple[float, float, float] = (0.7, 0.15, 0.15),
+):
+    """
+    Run Bayesian optimisation to estimate the position of the
+    sample image within the atlas volume.
+    Uses the registration objective function to optimise pitch, yaw, roll,
+    and z-slice for best similarity between the atlas and sample.
+
+    Parameters
+    ----------
+    atlas_volume : np.ndarray
+        3D atlas volume used as the reference.
+    sample : np.ndarray
+        2D image to be aligned to the atlas.
+    manual_z_range : tuple[float, float]
+        Lower and upper bounds for z-slice selection.
+        Defaults to None i.e. entire range.
+    pitch_bounds, yaw_bounds, roll_bounds : tuple[float, float], optional
+        Bounds for rotation angles (default: (-5, 5) degrees).
+    init_points : int, optional
+        Number of initial random points for Bayesian optimisation (default: 5).
+    n_iter : int, optional
+        Number of optimisation iterations (default: 15).
+    metric : Literal["mi", "ncc", "ssim", "combined"], optional
+        Similarity metric used to compare the sample and atlas slice:
+        - "mi"       : Mutual Information
+        - "ncc"      : Normalised Cross-Correlation
+        - "ssim"     : Structural Similarity Index
+        - "combined" : weights[0]*MI + weights[1]*NCC + weights[2]*SSIM
+        Defaults to "mi".
+    weights : tuple[float, float, float], optional
+        3-tuple specifying weights for (MI, NCC, SSIM) in the combined metric.
+        Only used if metric="combined". Must sum to 1.
+        Defaults to (0.7, 0.15, 0.15).
+
+    Returns
+    -------
+    dict
+        Dictionary containing the optimal alignment parameters:
+        - "best_pitch" : float
+            Optimal pitch angle (in degrees).
+        - "best_yaw" : float
+            Optimal yaw angle (in degrees).
+        - "best_roll" : float
+            Optimal roll angle (in degrees).
+        - "best_z_slice" : int
+            Optimal z-slice index within the rotated atlas.
+        - "done" : bool
+            Indicates that optimisation is complete (always True).
+
+    Prints
+    ------
+    Optimal parameters and similarity score for alignment.
+    """
+
+    if manual_z_range is None:
+        manual_z_range = (0, atlas_volume.shape[0] - 1)
+
+    def objective(pitch, yaw, z_slice):
+        return registration_objective(
+            pitch, yaw, 0, z_slice, atlas_volume, sample, metric, weights
+        )
+
+    pbounds = {
+        "pitch": pitch_bounds,
+        "yaw": yaw_bounds,
+        "z_slice": manual_z_range,
+    }
+
+    optimizer = BayesianOptimization(
+        f=None,
+        pbounds=pbounds,
+        verbose=2,
+        random_state=42,
+    )
+    # Customise Gaussian Progress
+    optimizer.set_gp_params(alpha=1e-3, n_restarts_optimizer=5)
+
+    # Initial random points
+    for _ in range(init_points):
+        point = optimizer.suggest()
+        score = objective(**point)
```

Comment:
If we rejig the arguments the arguments to `registration_objective` (get rid of `roll`) we could simplify this to:
```suggestion
        score = registration_objective(**point, atlas_volume, sample, metric, weights)
```

Reply:
If it complains about keyword arguments, you might have to do:
```suggestion
        score = registration_objective(**point, atlas_volume=atlas_volume, sample=sample, metric=metric, weights=weights)
```